staging/src/k8s.io/apiserver/pkg/admission/attributes.go

```diff
@@ -83,3 +93,54 @@ func (record *attributesRecord) GetOldObject() runtime.Object {
 func (record *attributesRecord) GetUserInfo() user.Info {
 	return record.userInfo
 }
+
+func (record *attributesRecord) GetAnnotations() map[string]string {
+	record.annotationLock.RLock()
+	defer record.annotationLock.RUnlock()
+
+	if record.annotations == nil {
+		return nil
+	}
+	cp := make(map[string]string)
+	for key, value := range record.annotations {
+		cp[key] = value
+	}
+	return cp
+}
+
+func (record *attributesRecord) SetAnnotations(annotations map[string]string) {
+	record.annotationLock.Lock()
+	defer record.annotationLock.Unlock()
+
+	for key, value := range annotations {
+		record.setAnnotation(key, value, false)
+	}
+}
+
+func (record *attributesRecord) SetAnnotation(key, value string) {
+	record.setAnnotation(key, value, true)
+}
+
+func (record *attributesRecord) setAnnotation(key, value string, lock bool) {
+	parts := strings.Split(key, "/")
+	if len(parts) != 2 {
+		glog.Warningf("Annotation key has invalid format, the right format is a DNS subdomain prefix and '/' and key name. (e.g. 'podsecuritypolicy.admission.k8s.io/admit-policy')")
+		return
+	}
+	if msgs := validation.IsQualifiedName(key); len(msgs) != 0 {
+		glog.Warningf("Annotation key has invalid format %s. A qualified name like 'podsecuritypolicy.admission.k8s.io/admit-policy' is required.", strings.Join(msgs, ","))
+		return
+	}
+
+	if lock {
+		record.annotationLock.Lock()
+		defer record.annotationLock.Unlock()
+	}
+	if record.annotations == nil {
+		record.annotations = make(map[string]string)
+	}
+	if v, ok := record.annotations[key]; ok && v != value {
+		glog.V(5).Infof("admission annotation overwrite, key:%q, old value: %q, new value:%q", key, record.annotations[key], value)
```

Comment:
what's this for? Only for debug? Then should delete this if

Reply:
We only log it when one key-value pair is overwroten.